### ISSUE TITLE
MEFLing causes bimodality around zero fluorescence

### ISSUE THREAD

**Issue by ppreshant** [taborlab/FlowCal#359]
I noticed that all of my samples have bimodal peaks around 0 fluorescence and I traced this to the MEFLing step of FlowCal in my workflow. This is most prominently visible in the negative control with no fluorescence shown below. 

Has this been seen before and would this be due to using old beads _which have a significant zero fluorescent beads_? I also see the same peak near zero fluorescence for the SEA beads in the Tabor lab (_which should be un-expired_) in the first amond 8 peaks.

Any hunch causing this would be appreciated :)

## Bimodal around 0

![S063c_Ds_empty-MEFLing](https://user-images.githubusercontent.com/14856479/223221775-feb8f4a7-f12a-4b76-bab2-0a6035045a0e.png)
Zoom in of the histogram post MEFLing plotted with 200 bins : <img src ="https://user-images.githubusercontent.com/14856479/223221830-6fbc8b4c-9be0-4899-a96f-e8dc14ec774e.png" width=250>

## Here's the beads processing graphs - 

![S063c_beads_gating](https://user-images.githubusercontent.com/14856479/223223405-c4355988-6dcb-4423-960d-54a74bdbd772.png)

![S063c_beads_MEFL](https://user-images.githubusercontent.com/14856479/223223417-2ccad67e-bf55-4228-b24b-cea73970ee9a.png)

![S063c_beads_fitting](https://user-images.githubusercontent.com/14856479/223223437-290fef87-78b8-4092-80d9-be418d3d523c.png)

**Comment by JS3xton** [comment]
Hey Prashant, my first guess is that this is a visualization artifact resulting from overbinning the data near zero. (I've seen more egregious versions of this in other contexts.) Do you know how the bin edges are being calculated for the post-calibration data? If the calibration function is scaling data away from zero and the bin edges are not similarly adjusted, spreading fewer data values over the near-zero bins can introduce this artificial bimodality.

If you're relying on the `logicle` scale defaults, the `FlowCal.plot.hist1d()` function should get the bin edges dynamically from the `FCSData` object, which will use the data and the FlowCal `_LogicleTransform` to generate bin edges ([here](https://github.com/taborlab/FlowCal/blob/v1.3.0/FlowCal/plot.py#L1215) -> [here](https://github.com/taborlab/FlowCal/blob/v1.3.0/FlowCal/io.py#L1573)). I think the resulting bin edges may depend on whether the data used are calibrated or uncalibrated, though, as this may change how the `_LogicleTransform` calculates the default `T`, `M`, and `W` Logicle scale parameters (see [pubmed 16604519](https://pubmed.ncbi.nlm.nih.gov/16604519/) for more on the Logicle scale and its parameters). If you're explicitly reusing the default _uncalibrated_ bin edges, I would be less surprised to see this kind of visualization artifact; if you're implicitly relying on the `FlowCal.plot.hist1d()` defaults (which should generate new, updated bin edges for the calibrated data), I'd have to walk more closely through an example to see how the `_LogicleTransform` is making the bin edges to determine if this might still be an overbinning issue.

(It would also be helpful to inspect the calibration model parameters and make sure they make sense--the exponential term should be near 1.0 for modern cytometers, and the linear scaling factor should depend on the detector voltage.)

**Comment by ppreshant** [comment]
Awesome, thank you! I will inspect the bin edges and the calibration parameters and get back to you.

I was calling the plot with this `FlowCal.plot.hist1d(processed_single_fcs, channel = fluorescence_channels[0], xlim = (-500, 500), bins = 200)`

**Comment by ppreshant** [comment]
I looked into this and it definitely has to do more with the calibration model parameters than the plotting.
This is a different dataset and different beads but here's some scatter2d plots showing the zero avoidance in the MEF outputs more clearly
- Raw data vs MEF transformed data
<img src = "https://user-images.githubusercontent.com/14856479/231315794-5b1a20bd-4f98-41ef-80b9-aa63210ba3d8.png" width=250> <img src = "https://user-images.githubusercontent.com/14856479/231315874-521cfabc-8a9f-4de7-8210-6b2bbced4ec6.png" width = 250>

Running `FlowCal.mef.get_transform_fxn()` with `full_output = True` and looking at the `.fitting` section gave these details about the calibration model
```
'beads_params': array(6.84108707e-01, 4.13784424e+00, 3.43116109e+03),
'beads_model_str': 'm*log(fl_rfi) + b = log(fl_mef_auto + fl_mef)'
```
 where m (slope) = 0.684 -- **is this the exponential term you mention that should be close to 1?**
 b (intercept) = 4.137
 fl_mef_auto (autofluorescence) = 3.43e3

Do these values look alright or is there somewhere else I should be looking / should I change my beads or something?
Thanks a lot for the help!

**Comment by castillohair** [comment]
This is very strange. I can think of a few explanations but it would be easier if you could send us a few .fcs files including a problematic sample and beads, along with the code you're using that results in this issue.

**Comment by ppreshant** [comment]
Awesome, thank you both for looking into this! 

## If you want to reproduce my processing exactly 
I have the data and the script all in this github repo : https://github.com/ppreshant/flow_cytometry 
I hope it will be reproducible for you, but all you need to do is run the jupyter notebook : `scripts_archive/`[adhoc_flowcal_analysis.ipynb](https://github.com/ppreshant/flow_cytometry/blob/main/scripts_archive/adhoc_flowcal_analysis.ipynb) until the `Explore MEFL calibration prameters` section. 
_I assume you would have all the flowcal dependancies or you could re-create my conda env in the repo using `conda env create -f flowcal_wrappers_environment.yaml`_
I wrote a lot of wrappers to flowcal functions for automated running on every dataset so it might be hard to follow what each function is doing. 


## alternatively
If you don't want to clone the repo, you could just run these [.fcs data](https://github.com/ppreshant/flow_cytometry/tree/main/flowcyt_data/S067b1_143_ww/Samples_for_testing) through your flowcal. To approximately reproduce my processing, I do this to my data before MEFLing using this script in [g8_process_single_fcs_flowcal.py](https://github.com/ppreshant/flow_cytometry/blob/main/scripts_general_fns/g8_process_single_fcs_flowcal.py) and [g14_gating_functions.py](https://github.com/ppreshant/flow_cytometry/blob/main/scripts_general_fns/g14_gating_functions.py)

```python
# gate out saturated events - high and low
singlefcs_gate1 = FlowCal.gate.high_low(single_fcs, channels = scatter_channels)

# auto density gating : for 50% of cells
singlefcs_densitygate = FlowCal.gate.density2d(singlefcs_gate1,
                                       channels = scatter_channels,
                                       gate_fraction = density_gating_fraction,
                                       full_output=True) # full => saves outline

# Singlets: auto density gating : for 90% of cells w FSC-A vs H
singlefcs_singlets90 = FlowCal.gate.density2d(singlefcs_densitygate.gated_data,
                                       channels = singlet_channels,
                                       gate_fraction = 0.90,
                                       full_output=True) # full => saves outline

# convert data into MEFLs
calibrated_fcs = beads_calibration_function(singlefcs_singlets90.gated_data, fluorescence_channels)
```
There wasn't one problem .fcs file but every sample which has no fluorescence does this bimodal thing around 0 after MEFLing
Here are some plots which give an overview of what MEFLing does to these cells without red fluorescence : _raw data on left and MEFLed on right_ (made through other scripts in R) 
- Also if it helps for context, I'm using the Sony SA3800 spectral flow cytometer that removes autofluorescence while doing automatic compensation (which is why the no fluorescence samples are very close to 0)

<img src="https://user-images.githubusercontent.com/14856479/231652604-d5842ad9-034a-4e76-8bd1-0e11f2c3e7f8.png" width = 300> <img src = "https://user-images.githubusercontent.com/14856479/231652886-6405f5ec-22b9-4cca-ad19-7e51a706c7a7.png" width = 300>

**Comment by ppreshant** [comment]
> This is very strange. I can think of a few explanations but it would be easier if you could send us a few .fcs files including a problematic sample and beads, along with the code you're using that results in this issue.

Were you able to check out the files or code I sent? If not could you give some hints of the explanations you were thinking of?

**Comment by castillohair** [comment]
I'm very sorry about the delay. I'm out of the office with family stuff until next Wednesday. I'll look into it when I come back. 

About the explanations: I guess I am puzzled by the slope of the standard curve, despite the plot looking relatively nice, so I was thinking about looking at that first. Not sure that would explain the dip at zero though. Sorry this is too vague, this is why I wanted to look at it.

I'll get to it as soon as I go back. Sorry again.

**Comment by ppreshant** [comment]
No worries. Please take your time.

I'll look into the standard curve slope you mention in the meantime.

**Comment by castillohair** [comment]
I looked at the data and I have some sense of what's going on. In summary, some of this is a result of how we chose to extend the standard curve to deal with "negative" fluorescence values product of the automatic background subtraction modern cytometers like to make (kids these days smh). But I also found that a lot of this can be mitigated with good beads processing. @JS3xton I think you will be interested in the second part that deals with the standard curve.

# Part 1: reproducing the problem with @ppreshant 's data

First, I processed @ppreshant 's data using a modification of one of our example scripts. Code is shown below for reference but I'll summarize the important points after so you don't have to read the code.
```python

import os
import os.path
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt
import FlowCal

###
# Definition of constants
###

beads_filename     = './Samples_for_testing/beads/A12 Well - A12 WLSM.fcs'

# Names of the FCS files containing data from cell samples
samples_filenames = [
    './Samples_for_testing/A01 Well - A01 WLSM.fcs',
    './Samples_for_testing/A02 Well - A02 WLSM.fcs',
    './Samples_for_testing/A03 Well - A03 WLSM.fcs',
    './Samples_for_testing/A04 Well - A04 WLSM.fcs',
    './Samples_for_testing/A06 Well - A06 WLSM.fcs',
    './Samples_for_testing/A07 Well - A07 WLSM.fcs',
    './Samples_for_testing/A08 Well - A08 WLSM.fcs',
    './Samples_for_testing/A09 Well - A09 WLSM.fcs',

]

# Fluorescence values of each bead subpopulation, in MEF
mefl_values     = np.array([0, 487, 1474, 4516, 11260, 34341, 107608, 260461])

# Plots will be generated at various stages of analysis. The following are the
# names of the folders in which we will store these plots.
beads_plot_dir   = 'plot_beads'
samples_plot_dir = 'plot_samples'

if __name__ == "__main__":

    # Check that plot directories exist, create if they do not.
    if not os.path.exists(beads_plot_dir):
        os.makedirs(beads_plot_dir)
    if not os.path.exists(samples_plot_dir):
        os.makedirs(samples_plot_dir)

    ###
    # Part 1: Processing calibration beads data
    ###
    print("\nProcessing calibration beads...")

    # Load calibration beads data from the corresponding FCS file.
    print("Loading file \"{}\"...".format(beads_filename))
    beads_sample     = FlowCal.io.FCSData(beads_filename)

    print("Performing data transformation...")
    beads_sample     = FlowCal.transform.to_rfi(beads_sample)

    # Gating
    print("Performing gating...")

    beads_sample_gated = FlowCal.gate.high_low(beads_sample,
                                               channels=['FSC-A'],
                                               low=(5000))

    density_gate_output = FlowCal.gate.density2d(
        data=beads_sample_gated,
        channels=['FSC-A','SSC-A'],
        gate_fraction=0.8,
        sigma=5.,
        full_output=True)
    beads_sample_gated = density_gate_output.gated_data
    gate_contour       = density_gate_output.contour

    # Plot forward/side scatter 2D density plot and 1D fluorescence histograms
    print("Plotting density plot and histogram...")

    # Parameters for the forward/side scatter density plot
    density_params = {}
    density_params['mode'] = 'scatter'
    density_params['sigma'] = 5.

    # Plot filename
    plot_filename     = '{}/density_hist_{}.png'.format(beads_plot_dir,
                                                        'beads')

    # Plot and save
    FlowCal.plot.density_and_hist(
        beads_sample,
        beads_sample_gated,
        density_channels=['FSC-A', 'SSC-A'],
        hist_channels=['mScarlet-I-A'],
        gate_contour=gate_contour,
        density_params=density_params,
        savefig=plot_filename)

    # Use beads data to obtain a MEF transformation function
    print("\nCalculating standard curve for channel mScarlet-I-A...")

    mef_transform_fxn = FlowCal.mef.get_transform_fxn(
        beads_sample_gated,
        mef_channels='mScarlet-I-A',
        mef_values=mefl_values,
        # clustering_channels=['FL1', 'FL3'],
        verbose=True,
        plot=True,
        plot_dir=beads_plot_dir,
        plot_filename='beads')
    
    ###
    # Part 2: Processing cell sample data
    ###
    print("\nProcessing cell samples...")

    # We will use the list ``samples`` to store processed, transformed flow
    # cytometry data of each sample.
    samples = []

    # Iterate over cell sample filenames
    for sample_id, sample_filename in enumerate(samples_filenames):

        # Load sample data from the corresponding FCS file.
        print("\nLoading file \"{}\"...".format(sample_filename))
        sample = FlowCal.io.FCSData(sample_filename)
        
        # Perform transformations
        print("Performing data transformation...")

        # Transform data to RFI, as done previously with beads.
        sample = FlowCal.transform.to_rfi(sample)

        # First process data without calibration
        # Gating
        print("Performing gating on non-calibrated data...")

        # Apply density gating on the forward/side scatter channels.
        # Return also a contour around the gated region.
        density_gate_output = FlowCal.gate.density2d(
            data=sample,
            channels=['FSC-A','SSC-A'],
            gate_fraction=0.90,
            full_output=True)
        sample_gated = density_gate_output.gated_data
        gate_contour = density_gate_output.contour

        # Plot forward/side scatter 2D density plot and 1D fluorescence
        # histograms
        print("Plotting density plot and histogram...")

        # Parameters for the forward/side scatter density plot
        density_params = {}
        density_params['mode'] = 'scatter'

        # Parameters for the fluorescence histograms
        hist_params = {}
        hist_params['xlabel'] = 'mScarlet-I-A ' + \
            '(a.u.)'

        # Plot filename
        plot_filename = '{}/density_hist_{}.png'.format(
            samples_plot_dir,
            'S{:03}'.format(sample_id + 1))

        # Plot and save
        FlowCal.plot.density_and_hist(
            sample,
            sample_gated,
            density_channels=['FSC-A','SSC-A'],
            hist_channels=['mScarlet-I-A'],
            gate_contour=gate_contour,
            density_params=density_params,
            hist_params=hist_params,
            savefig=plot_filename)

        # We now use the transformation function obtained from calibration beads
        # to transform FL1 data from RFI to MEF.
        sample_mef = mef_transform_fxn(sample, channels=['mScarlet-I-A'])

        # Gating
        print("Performing gating on calibrated data...")

        # Apply density gating on the forward/side scatter channels.
        # Return also a contour around the gated region.
        density_gate_output = FlowCal.gate.density2d(
            data=sample_mef,
            channels=['FSC-A','SSC-A'],
            gate_fraction=0.90,
            full_output=True)
        sample_mef_gated = density_gate_output.gated_data
        gate_mef_contour = density_gate_output.contour

        # Plot forward/side scatter 2D density plot and 1D fluorescence
        # histograms
        print("Plotting density plot and histogram...")

        # Parameters for the forward/side scatter density plot
        density_params = {}
        density_params['mode'] = 'scatter'

        # Parameters for the fluorescence histograms
        hist_params = {}
        hist_params['xlabel'] = 'mScarlet-I-A ' + \
            '(MEF)'

        # Plot filename
        # The figure can be saved in any format supported by matplotlib (svg,
        # jpg, etc.) by just changing the extension.
        plot_filename = '{}/density_hist_{}_mef.png'.format(
            samples_plot_dir,
            'S{:03}'.format(sample_id + 1))

        # Plot and save
        # In this case, we will plot the forward/side scatter channels in
        # the density plot and a histogram of the fluorescence channel FL1
        # below.
        FlowCal.plot.density_and_hist(
            sample_mef,
            sample_mef_gated,
            density_channels=['FSC-A','SSC-A'],
            hist_channels=['mScarlet-I-A'],
            gate_contour=gate_mef_contour,
            density_params=density_params,
            hist_params=hist_params,
            savefig=plot_filename)
```

While doing this, I found that
1) calling `gate.high_low()` with default parameters (i.e. without explicit high or low parameters) is not a good idea for your data. @JS3xton : this function clips by default by taking `low` and `high` from `FCSData.range`, which is initialized to `[0, $PnR]`, but this is not appropriate for floating-point .fcs data with negative values. `gate.high_low()` actually uses -inf if `range[0]` is `None`, but from a quick glance I don't think `FCSData` sets `range` intelligently if the data type is floating point. We might wanna look into this more.
2) I don't think `gate.start_end()` is necessary either.

3) your beads data has A LOT of debris which made it really hard to fit a standard curve. I had success via `gate.high_low()` with `low=5000` before any other processing, which removed most of the debris and allowed me to call density gating with a high gate fraction.

After this, the beads parameters I got were `[  0.92   1.2  112.69]`. Note that the slope (first parameter) is much closer to one than what you mentioned @ppreshant (~0.68).

Still, I did see the double peak around zero in cell samples with vs. without calibration.

Without calibration:
![density_hist_S001](https://user-images.githubusercontent.com/1529481/236641115-c8cbabc0-36ee-4bcc-b1c4-74381bce1f60.png)

With calibration:
![density_hist_S001_mef](https://user-images.githubusercontent.com/1529481/236641126-aff43817-82e0-4e4b-8ce3-68951f988b96.png)

# Part 2: looking at the standard curve function

Because all the previous processing looked good to me, I moved on to the standard curve itself, trying to figure out if there was some weird numerical error. If you look at the MEF module documentation you will find that MEF fluorescence is calculated as

```
s(fl_rfi) = sign(fl_rfi) * exp(b) * (abs(fl_rfi)**m)
```

This corresponds to the red line in the standard curve plot, with a few additions. The `abs()` and the `sign()` functions were added to extend the function to negative values, since calibration is done only with positive fluorescence values. This, combined with `m!=1` can create some weird artifacts close to zero.

We can illustrate this with simulated data:

```python
import numpy
import matplotlib
from matplotlib import pyplot

def sc_func(x):
    return numpy.sign(x) * numpy.exp(p[1]) * (numpy.abs(x)**p[0])

# Fluorescence model parameters
p = [  0.68,   1.2,  112.69]

# Simulate converting au to mef
x = numpy.concatenate((
    -numpy.logspace(-8, 4, 1000)[::-1],
    [0],
    numpy.logspace(-8, 4, 1000),
))
y = sc_func(x)

fig, ax = pyplot.subplots()
ax.plot(x, y)
ax.set_xlabel('"a.u."')
ax.set_ylabel('"MEF"')

xlim = 1000
ylim = (xlim**p[0])*numpy.exp(p[1])
ax.set_xlim(-xlim, xlim)
ax.set_ylim(-ylim, ylim)
```

![image](https://user-images.githubusercontent.com/1529481/236641406-9ac6cb91-cfd5-4e2f-ab30-b630a68ff637.png)

Because `m < 1`, values close to zero are pulled farther away from zero (for example, sqrt(x) = x^0.5 > x if x is a small number). This is more evident if we simulate a distribution

```python
# Histogram test
# Generate normally distributed x around zero
# See what happens with the mef-transformed distribution
xn = numpy.random.normal(scale=1000, size=100000)
yn = sc_func(xn)

fig, axes = pyplot.subplots(2, 1, figsize=(8, 4))

ax = axes[0]
ax.hist(xn, bins=100)
ax.set_ylabel('"a.u."')

ax = axes[1]
ax.hist(yn, bins=100)
ax.set_ylabel('"MEF"')
```
![image](https://user-images.githubusercontent.com/1529481/236641480-16b4c753-b0a7-4869-831a-bc4fa2a4fc29.png)

So we can reproduce this issue with just simulated data.

It should be obvious that the closer `m` is to one, the less dramatic this effect will be. If we repeat this with `m=0.92`:

![image](https://user-images.githubusercontent.com/1529481/236641527-776f4de3-1936-4f4c-9eef-aadaaa8b5e60.png)

And with `m > 1` we have the opposite effect. For `m = 1.1`:

![image](https://user-images.githubusercontent.com/1529481/236641548-d9c69a0a-a0f4-4d0f-941e-38c781daa10f.png)

So now the question is what to do with this? I think these are your options:
- First, make sure that your beads processing is correct. `m=0.68` is highly atypical and suspicious. Ideally you would have it between 0.9 - 1.1, and these should be extremely consistent across your runs (i.e. if you got 0.92 once, you should get +/- 0.1 at most every time in the same instrument). Gating out debris more aggressively, as I did above, might help. Cleaning and doing maintenance on your instrument so the debris is reduced in the first place might also help a lot. Also make sure you're using the right MEF numbers for your channel.
- Now, if `m` is not exactly one you will still have artifacts around zero to some extent. `m` is  **really** close to one in most instruments, but we had `m=1.08` in our really old cytometer in the Tabor lab so we know large deviations can happen (although as I mentioned these should be consistent).
  - One option is just to ignore the remaining artifacts. To be honest I doubt this will affect most analysis that I can think of.
  - The other option is to force `m=1`. We made it a floating parameter because it made for a better fit, but if these artifacts bother you a lot you could just force it to be one and live with a slightly worse standard curve fit. I think you could easily do this in a hacky way by removing `p[0]` from line 451 in the FlowCal `mef` module, function `fit_beads_autofluorescence()`, inner function `sc_fun()`. If you do this and have trouble I wouldn't be able to provide help since it's a modification I never tried myself.

Hope this helps

**Comment by ppreshant** [comment]
Thanks at lot for the elaborate response @castillohair! applying the MEFL function to a random distribution makes it very clear.

I will definitely make sure that m is in the range `0.9 - 1.1` you mention. I tried increasing the gating threshold in the instrument for the beads to remove some the of the debris this time, and will do the `gate.high_low()` more aggressively too. Thanks for taking the time :)

**Comment by castillohair** [comment]
Glad it helped!

I'm closing this issue now.

**Comment by JS3xton** [comment]
Nice demonstration, @castillohair.

Yeah, I've also seen `FCSData.range` be unreliable when the data type is floating point. I'd be open to reducing our reliance on `range`, or at least testing the data type before using it. Not sure to what extent inaccurate `range` values compromise other things like `FCSData.hist_bins()`.

I've actually re-implemented calibration in another context, and I simplified the calibration function by hardcoding `m` as 1. This was simpler and yielded some nice math simplifications. I'd be open to making this the FlowCal default, since modern instruments generally conform to this. We'd have to think about how/if we want to support a function where `m` is not one, though, for backwards compatibility.
